### PR TITLE
GEODE-9016: Fix the NPE for PutAll with CQ LOCAL_DESTROY message type

### PR DESCRIPTION
**Pull request by jchen21**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/DistributedPutAllOperation.java [hunk 1]
```diff
@@ -814,6 +818,46 @@ protected FilterRoutingInfo getRecipientFilterRouting(Set cacheOpRecipients) {
...
+      CacheDistributionAdvisor.CacheProfile cf =
+          (CacheDistributionAdvisor.CacheProfile) ((Bucket) getRegion()).getPartitionedRegion()
+              .getCacheDistributionAdvisor().getProfile(m);
+
+      if (cf == null || cf.filterProfile == null || cf.filterProfile.isLocalProfile()
+          || cf.filterProfile.getCqMap().isEmpty()) {
+        continue;
+      }
+
+      for (Object value : cf.filterProfile.getCqMap().values()) {
```
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/cache/DistributedPutAllOperation.java:838]: In the future, you might want to try extracting some of those inner-loops or inner-if-blocks to their own methods. How you name those methods can then really improve understanding. It's also too easy to get lost in those deeply nested structures like this.

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java [hunk 2]
```diff
@@ -120,6 +123,60 @@
...
+      }
+      region.putAll(buffer);
+    });
+
+    client.invoke(() -> {
+      QueryService cqService = getCache().getQueryService();
+      CqQuery cqQuery = cqService.getCq(cqName);
+      if (cqQuery == null) {
+        fail("Failed to get CQ " + cqName);
+      }
```
Comment (kirklund) [inline geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:157]: Please use assertions instead of if-fail patterns:
```
assertThat(cqQuery)
    .withFailMessage("Failed to get CQ " + cqName)
    .isNotNull();
```

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java [hunk 3]
```diff
@@ -120,6 +123,60 @@
...
+      try {
+        cqQuery.executeWithInitialResults();
+      } catch (Exception ex) {
+        fail("Failed to execute  CQ " + cqName, ex);
+      }
```
Comment (kirklund) [inline geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:162]: You don't need to catch Exception and invoke fail. It's better to let `VM.invoke(...)` throw the Exception wrapped in an RMIException. You already have `throws Exception` on the test method so JUnit will handle it and display the full stack perfectly:
```
cqQuery.executeWithInitialResults();
```

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java [hunk 4]
```diff
@@ -120,6 +123,60 @@
...
+      for (int i = 1; i < numObjects; i++) {
+        Portfolio p = new Portfolio(-1 * i);
+        buffer.put("" + i, p);
+      }
+      region.putAll(buffer);
+    });
+
+    cqHelper.closeClient(client);
+    cqHelper.closeServer(server2);
+    cqHelper.closeServer(server1);
```
Comment (kirklund) [inline geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:177]: It's better (and more industry standard for writing JUnit tests) if you can figure out a way to move cleanup code at the end of the test method to a `tearDown` method annotated with `@After`. For example, if you make `cqHelper.closeClient` return without error if client is null or not running, then you could hoist these local variables to fields and the test would look more like this:
```
private VM client;
private VM server1;
private VM server2;

@Before 
public void setUp() {
 server1 = getVM(0);
 server2 = getVM(1);
 client = getVM(2);
}

@After
public void tearDown() {
  cqHelper.closeClient(client);
  cqHelper.closeServer(server2);
  cqHelper.closeServer(server1);
}
```
And `cqHelper` close methods could have some early-outs like this:
```
public void closeClient(VM vm) {
  if (vm == null) {
    return;
  }
  // etc
}
```

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java [hunk 5]
```diff
@@ -120,6 +123,60 @@
...
+    final String host = Host.getHost(0).getHostName();
+    final int port = server2.invoke(() -> PartitionedRegionCqQueryDUnitTest.getCacheServerPort());
+    createClient(client, port, host);
+    createCQ(client, cqName, cqs[0]);
+
+    int numObjects = 1000;
+
+    server1.invoke(() -> {
+      Region region = getCache().getRegion(SEPARATOR + "root" + SEPARATOR + regions[0]);
+      Map buffer = new HashMap();
```
Comment (kirklund) [inline geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:144]: You should avoid raw types in both product and test code. For example: `Region<String, Portfolio>` or even `Region<String, Object>`. Same with Map.

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java [hunk 6]
```diff
@@ -120,6 +123,60 @@
...
+  @Test
+  public void testPutAllWithCQLocalDestroy() throws Exception {
+    VM server1 = getVM(0);
+    VM server2 = getVM(1);
+    VM client = getVM(2);
+
+    final String cqName = "testPutAllWithCQLocalDestroy_0";
+    createServer(server1);
+    createServer(server2);
+    final String host = Host.getHost(0).getHostName();
```
Comment (kirklund) [inline geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:135]: Avoid `Host` and just use:
```
String hostName = VM.getHostName();
```

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java [hunk 7]
```diff
@@ -120,6 +123,60 @@
...
+    final int port = server2.invoke(() -> PartitionedRegionCqQueryDUnitTest.getCacheServerPort());
```
Comment (kirklund) [inline geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:136]: Remove the name of the test anytime you see this, so you just have:
```
int port = server2.invoke(() -> getCacheServerPort());
```

**Review by kirklund — CHANGES_REQUESTED** [review]
I marked some areas in the dunit test that could be improved. My only other recommendation is to please figure out a way to add test(s) to the existing (but mostly empty) DistributedPutAllOperationTest unit test using mocks for at least the changes involving `removeDestroyTokensFromCqResultKeys`.

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java (hunk 5 above)
Comment (kirklund) [reply geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:144]: PS: You don't really need `SEPARATOR + "root" + SEPARATOR`. This will also work:
```
getCache().getRegion(regions[0]);
```

File: geode-core/src/main/java/org/apache/geode/internal/cache/DistributedPutAllOperation.java [hunk 8]
```diff
@@ -814,6 +818,46 @@ protected FilterRoutingInfo getRecipientFilterRouting(Set cacheOpRecipients) {
     return consolidated;
   }
 
+  @Override
+  protected void removeDestroyTokensFromCqResultKeys(FilterRoutingInfo filterRouting) {
```
Comment (agingade) [inline geode-core/src/main/java/org/apache/geode/internal/cache/DistributedPutAllOperation.java:822]: The same may need to be done in RemoveAll.

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java [hunk 9]
```diff
@@ -120,6 +123,60 @@
 
   private static int bridgeServerPort;
 
+  @Test
+  public void testPutAllWithCQLocalDestroy() throws Exception {
```
Comment (agingade) [inline geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:127]: Test needs to be added for removeAll.

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java (hunk 5 above)
Comment (jchen21) [reply geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:144]: Actually, for this test, we still need `SEPARATOR + "root" + SEPARATOR`, otherwise `getRegion()` can't find the region.

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java (hunk 4 above)
Comment (jchen21) [reply geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:177]: I would love the clean the test code as you suggested. This test also uses a lot of already deprecated test classes and APIs. It's better to refactor it. However, introducing @Before and @After would affect all 14 test methods in the file, it would be better to have a separate JIRA to track the refactoring. I have created https://issues.apache.org/jira/browse/GEODE-9020 to track it.

Comment (kirklund) [reply geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:177]: Feel free to fit into the test in the best way you can. Another alternative is to extract the new test out to a new test class (now in a followup PR). You could also modernize the whole test but that would be best done as a followup PR.

**Review by kirklund — APPROVED** [review]

File: geode-core/src/test/java/org/apache/geode/internal/cache/DistributedPutAllOperationTest.java [hunk 10]
```diff
@@ -33,4 +47,39 @@ public void shouldBeMockable() throws Exception {
 
     assertThat(mockDistributedPutAllOperation.getBaseEvent()).isSameAs(mockEntryEventImpl);
   }
+
+  @Test
+  public void testDoRemoveDestroyTokensFromCqResultKeys() {
```
Comment (agingade) [inline geode-core/src/test/java/org/apache/geode/internal/cache/DistributedPutAllOperationTest.java:52]: A small refactoring request...to make it readable, can this be broken down (with new lines) into:
setup
action
validation/verification

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java [hunk 11]
```diff
@@ -120,6 +125,109 @@
...
+
+    final String cqName = "testPutAllWithCQLocalDestroy_0";
+    createServer(server1);
+    createServer(server2);
+    final String host = VM.getHostName();
+    final int port = server2.invoke(() -> getCacheServerPort());
+    createClient(client, port, host);
+    createCQ(client, cqName, cqs[0]);
+
+    int numObjects = 1000;
```
Comment (agingade) [inline geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:142]: Does this need to be that big...If its not needed can we keep it small, say 5 or 10. Makes test to run faster.

**Review by agingade — APPROVED** [review]
Approving with the comments taken care.

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java (hunk 11 above)
Comment (jchen21) [reply geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:142]: Yes. It does need a larger number. Small number sometimes does not consistently reproduce the bug.

**Review by DonalEvans — APPROVED** [review]

File: geode-core/src/main/java/org/apache/geode/internal/cache/DistributedPutAllOperation.java [hunk 12]
```diff
@@ -814,6 +817,23 @@ protected FilterRoutingInfo getRecipientFilterRouting(Set cacheOpRecipients) {
...
+        @Unretained
+        EntryEventImpl entryEvent = getEventForPosition(i);
+        if (entryEvent != null && entryEvent.getKey() != null && cq != null
+            && cq.getFilterID() != null && cq.getFilterID().equals(cqID)
+            && e.getValue() != null && e.getValue().equals(MessageType.LOCAL_DESTROY)) {
+          cq.removeFromCqResultKeys(entryEvent.getKey(), true);
+        }
+      }
+    }
+  }
```
Comment (nabarunnag) [inline geode-core/src/main/java/org/apache/geode/internal/cache/DistributedPutAllOperation.java:836]:
```suggestion
  void doRemoveDestroyTokensFromCqResultKeys(FilterInfo filterInfo, ServerCQ cq, int length) {
    for (Map.Entry<Long, Integer> e : filterInfo.getCQs().entrySet()) {
      Long cqID = e.getKey();
      // For the CQs satisfying the event with destroy CQEvent, remove
      // the entry from CQ cache.
      for (int i = 0; i < length; i++) {
        @Unretained
        EntryEventImpl entryEvent = getEventForPosition(i);
        if (entryEvent != null && entryEvent.getKey() != null && cq != null
            && cq.getFilterID() != null && cq.getFilterID().equals(cqID)
            && e.getValue() != null && e.getValue().equals(MessageType.LOCAL_DESTROY)) {
          cq.removeFromCqResultKeys(entryEvent.getKey(), true);
        }
      }
    }
  }
```

Comment (nabarunnag) [reply geode-core/src/main/java/org/apache/geode/internal/cache/DistributedPutAllOperation.java:836]: Maybe this method can be put in a common class and reused in DistributedRemoveAllOperation and DistributedPutAllOperations

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java [hunk 13]
```diff
@@ -120,6 +125,111 @@
...
+          getCache().getRegion(SEPARATOR + "root" + SEPARATOR + regions[0]);
+      // PutAll with entries that do not satisfy CQ. This is to generate LOCAL_DESTROY CQ event
+      Map<String, Object> buffer = new HashMap();
+      for (int i = 1; i < numObjects; i++) {
+        Portfolio p = new Portfolio(-1 * i);
+        buffer.put("" + i, p);
+      }
+      region.putAll(buffer);
+    });
+
```
Comment (nabarunnag) [inline geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:175]: Should there be an assert at the end of the test to verify the behavior ?

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java [hunk 14]
```diff
@@ -120,6 +125,111 @@
...
+      Region<String, Object> region =
+          getCache().getRegion(SEPARATOR + "root" + SEPARATOR + regions[0]);
+      Set<String> keys = new HashSet<>();
+      for (int i = 1; i < numObjects; i++) {
+        keys.add("" + i);
+      }
+      // This is to generate LOCAL_DESTROY CQ event
+      region.removeAll(keys);
+    });
+
```
Comment (nabarunnag) [inline geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:227]: Should there be an assert at the end of the test to verify the behavior ?

**Review by nabarunnag — CHANGES_REQUESTED** [review]

File: geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java (hunk 11 above)
Comment (agingade) [reply geode-cq/src/distributedTest/java/org/apache/geode/cache/query/cq/dunit/PartitionedRegionCqQueryDUnitTest.java:142]: Can you please help me understand why consistency an issue here; in that case how sure we are that 1000 entries will recreate the issue all the time.

File: geode-core/src/main/java/org/apache/geode/internal/cache/DistributedPutAllOperation.java (hunk 12 above)
Comment (jchen21) [reply geode-core/src/main/java/org/apache/geode/internal/cache/DistributedPutAllOperation.java:836]: I don't think it is necessary to pass the length of `putAllData` as a parameter. `putAllData` is a field of `DistributedPutAllOperation`. We can easily get its length without getting it from a parameter. Maybe I am missing something you proposed.

For code refactoring with a common class of `DistributedRemoveAllOperation` and `DistributedPutAllOperation`, if necessary, it would be better to do it in another pull request. This pull request focuses on fixing the bug.

**Review by nabarunnag — APPROVED** [review]
LGTM
